### PR TITLE
update loss computation in modeling code

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

cc @stas00 

Follow up of #36607, see [this comment](https://github.com/huggingface/transformers/pull/36607#issuecomment-2781494656).

The following code snippet gives 

None

tensor(7.0538, grad_fn=<NllLossBackward0>)

This is particular necessary for context parallel to run correctly.

**Once the changes is approved, I will update all other places and add documentations.**

```python
import torch
from transformers import AutoModelForCausalLM

repo_id = "meta-llama/Llama-3.2-1B"
token = "YOUR_HF_TOKEN"

model = AutoModelForCausalLM.from_pretrained(repo_id, token=token)
input_ids = torch.ones(size=(1, 16), dtype=torch.int64)
labels = input_ids.clone()
shift_labels = torch.nn.functional.pad(labels[..., 1:], (0, 1), value=-100)
outputs_with_labels = model(input_ids, labels=labels)
outputs_with_shift_labels = model(input_ids, shift_labels=shift_labels)

print(outputs_with_labels.loss)
print(outputs_with_shift_labels.loss)
```

### CODE REVIEW

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -825,7 +825,7 @@ def forward(
         logits = self.lm_head(hidden_states[:, slice_indices, :])
 
         loss = None
-        if labels is not None:
+        if labels is not None or kwargs.get("shift_labels", None) is not None:
```
Comment (ArthurZucker) [inline src/transformers/models/llama/modeling_llama.py:828]: why not leave it to the loss function?

Comment (ydshieh) [reply src/transformers/models/llama/modeling_llama.py:828]: We can, but it means that we have to compute the loss outside the model's `forward` with 
the user have to do

```python
from transformers import ForCausalLMLoss

shift_labels = .... 
loss = ForCausalLMLoss(logits=logits, shift_labels=shift_labels, vocab_size=model.config.vocab_size)
```

It's not too much work for users, but it is nice if we could make it easier for them (they only need to take care of preparing `shift_labels `).

Enable this API means people could perform the same workflow with `labels` and with `shift_labels` (i.e. put them in model.forward and get it from outputs).

Comment (ydshieh) [reply src/transformers/models/llama/modeling_llama.py:828]: Also, we can't just pass `shift_labels` without passing `labels`

> loss = ForCausalLMLoss(logits=logits, shift_labels=shift_labels, vocab_size=model.config.vocab_size)

as it is a `required positional argument`. We have to do

> loss = ForCausalLMLoss(logits=logits, labels=None, shift_labels=shift_labels, vocab_size=model.config.vocab_size)

This kind of details could hide from user if we do it in modeling code.

Comment (stas00) [reply src/transformers/models/llama/modeling_llama.py:828]: yeah, someone was reported using a side-effect hack - passing `labels=garbage, shift_labels=real_data` and getting the `model` to give them the loss ;)

**Review by Muhtasham — APPROVED** [review]

**Comment by ydshieh** [comment]
@bot /style

**Comment by ydshieh** [comment]
@bot /style

File: src/transformers/models/llama/modeling_llama.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/llama/modeling_llama.py:707]: what I meant @ydshieh is that if we do this we need to update all models!

**Review by ArthurZucker — COMMENTED** [review]
Happy to have this merged, just needs:
- either all model update with make fix copies
- put this in the loss_function instead (not sure if possible)

**Comment by ydshieh** [comment]
> put this in the loss_function instead (not sure if possible)

This is unfortunately impossible, as in current modeling code, it only compute `loss` if `labels` is passed
```
if labels is not None:
    loss = ...
```
and if we want to use `shift_labels`, it won't be taken into account if we also give it `labels`, which is strange behavior.


I will go

> either all model update with make fix copies

🤞

**Comment by ydshieh** [comment]
Ready for an review 🙏 

Mostly, change

> if labels is not None:

to 

> if labels is not None or kwargs.get("shift_labels", None) is not None:

for `CausalLM` (or some `ForConditionalGeneration`) models.

**Comment by ydshieh** [comment]
I would love to change `shift_labels` to `shifted_labels` (also in `def ForCausalLMLoss`), as `shift_labels` sounds like a bool argument. Just nit however.

**Comment by ydshieh** [comment]
Yeah, I understand. No big deal, I guess probably people focus on make AI brrrr would ever care if there is `ed` or not 😅

**Review by ArthurZucker — COMMENTED** [review]
not a big fan of this, IDK if we have a better solution. WDYT about a flag saying they are already shifted? (it's just that kwargs are not to be used here but I mean.... yeah

**Comment by ydshieh** [comment]
IIRC, @ArthurZucker you mean we allow `labels` could be a shifted label, and we introduce a flag to indicate this, in

```
def ForCausalLMLoss(
```

right? Technically it's doable, but the new flag (say, `is_shifted`) will be overlapped with the function of `shift_labels`, which is not ideal. 


One possibility is we allow `shift_labels` to be `Optional[Union[torch.Tensor, bool]]`, then we do inside `ForCausalLMLoss`


```
    if shift_labels is True and labels is not None:
        shift_labels = labels
    ....
```

it could work and it could avoid the changes in the many modeling files. But you see the downside (same argument having 2 possible types and perform something differently).

@stas00 Any comment here?

**Comment by ydshieh** [comment]
@ArthurZucker @stas00 

This is probably a better and clean solution

https://github.com/huggingface/transformers/pull/38533/files

Let me know your opinions

**Comment by stas00** [comment]
@ydshieh, the last one is a smooth solution!
